opm/material/fluidsystems/blackoilpvt/Co2GasPvt.hpp

```diff
@@ -225,8 +268,85 @@ class Co2GasPvt
     Scalar gasReferenceDensity(unsigned regionIdx) const
     { return gasReferenceDensity_[regionIdx]; }
 
+    Scalar oilReferenceDensity(unsigned regionIdx) const
+    { return brineReferenceDensity_[regionIdx]; }
+
+    Scalar waterReferenceDensity(unsigned regionIdx) const
+    { return brineReferenceDensity_[regionIdx]; }
+
+    Scalar salinity(unsigned regionIdx) const
+    { return salinity_[regionIdx]; }
+
 private:
+
+    template <class LhsEval>
+    LhsEval rvwSat_(unsigned regionIdx,
+                    const LhsEval& temperature,
+                    const LhsEval& pressure) const
+    {
+        if (!enableVaporization_)
+            return 0.0;
+
+        // calulate the equilibrium composition for the given
+        // temperature and pressure.
+        LhsEval xgH2O;
+        LhsEval xlCO2;
+        BinaryCoeffBrineCO2::calculateMoleFractions(temperature,
+                                                    pressure,
+                                                    salinity_[regionIdx],
+                                                    /*knownPhaseIdx=*/-1,
+                                                    xlCO2,
+                                                    xgH2O,
+                                                    extrapolate);
+
+        // normalize the phase compositions
+        xgH2O = max(0.0, min(1.0, xgH2O));
+
+        return convertXgWToRvw(convertxgWToXgW(xgH2O), regionIdx);
+    }
+
+    /*!
+     * \brief Convert the mass fraction of the water component in the gas phase to the
+     *        corresponding water vaporization factor.
+     */
+    template <class LhsEval>
+    LhsEval convertXgWToRvw(const LhsEval& XgW, unsigned regionIdx) const
+    {
+        Scalar rho_wRef = brineReferenceDensity_[regionIdx];
+        Scalar rho_gRef = gasReferenceDensity_[regionIdx];
+
+        return XgW/(1.0 - XgW)*(rho_gRef/rho_wRef);
+    }
+
+    /*!
+     * \brief Convert a water vaporization factor to the the corresponding mass fraction
+     *        of the water component in the gas phase.
+     */
+    template <class LhsEval>
+    LhsEval convertRvwToXgW_(const LhsEval& Rvw, unsigned regionIdx) const
+    {
+        Scalar rho_wRef = brineReferenceDensity_[regionIdx];
+        Scalar rho_gRef = gasReferenceDensity_[regionIdx];
+
+        const LhsEval& rho_wG = Rvw*rho_gRef;
+        return rho_wG/(rho_wRef + rho_wG);
```

Comment:
```
const LhsEval& rho_wG = Rvw*rho_gRef;
return rho_wG/(rho_wRef + rho_wG);
```
Think the code above should be
```
const LhsEval& rho_wG = Rvw*rho_wRef;
return rho_wG/(rho_gRef + rho_wG);
```
Here is my derivation:
`Rvw = XgW/(1.0 - XgW)*(rho_gRef/rho_wRef)`  from  Line 318

`Rvw = XgW*rho_gRef/((1.0 - XgW)*rho_wRef)`

`(1.0 - XgW)*rho_wRef*Rvw = XgW*rho_gRef`

`rho_wRef*Rvw -  XgW*rho_wRef*Rvw = XgW*rho_gRef`

`XgW*rho_gRef +  XgW*rho_wRef*Rvw = rho_wRef*Rvw `

`XgW*(rho_gRef +  rho_wRef*Rvw) = rho_wRef*Rvw `

`XgW = rho_wRef*Rvw/(rho_gRef +  rho_wRef*Rvw) `

Reply:
Thanks for noticing. I will fix.